### ISSUE TITLE
SchemaPack: mechanism-linked world-model audit records from representation to behavior

### ISSUE THREAD

**Issue by crates** [entif-ai/rosetta#1633]
## 1. Summary

Create an experimental **Mechanistic World-Model Audit SchemaPack** and provider-neutral interoperability profile grounded in Pierre Beckmann, Matthieu Queloz, and André Freitas, [“World Modeling in Transformers”](https://arxiv.org/abs/2609.21748), its [project page](https://bepierre.github.io/world-modeling/), and the authors’ [MIT-licensed experiment suite](https://github.com/bepierre/world-modeling).

The paper shows that behavioral failure need not imply an incoherent internal world representation. TaxiGPT learns a highly faithful Manhattan graph, tracks position through a rolling history, and uses a two-dimensional goal compass, yet out-of-distribution stress weakens the current-position write and raises interference among superposed intersection features. The resulting failures can be classified as fatal slips, silent slips, full corruption, and give-up slips. Causal edits that restore the true-position signal or reduce interference recover much of the behavior.

The proposed profile preserves a non-collapse boundary among:

- an environment/data/model checkpoint;
- a learned representational feature or subspace;
- a decoded state estimate;
- a causal intervention and its target;
- an observed action distribution or trajectory;
- a failure-classification rule and result;
- a recovery evaluation;
- a bounded downstream reliance claim.

Rosetta is a free, open-source project, not intended for sale or monetization; it is intended to pursue universally applicable advances in AI that could benefit humanity if successfully realized. This proposal is strictly non-commercial and does not imply endorsement or collaboration by the paper’s authors.

## 2. Research basis

Primary research and implementation:

- Paper: https://arxiv.org/abs/2609.21748
- Project page: https://bepierre.github.io/world-modeling/
- Code: https://github.com/bepierre/world-modeling
- Code license: MIT
- Experiment index: https://github.com/bepierre/world-modeling/blob/main/experiments/README.md

Relevant Entif publications and specifications:

- [We Had the Seeds of a Babel Fish](https://entif.ai/tags/research/2026/09/10/babel-fish-alien-mind/)
- [Rosetta Pasigraphy Protocol](https://entif.ai/tags/research/2026/09/09/rosetta-pasigraphy-protocol/)
- [Rosetta v3.0.0 Core Spine Specification](https://github.com/entif-ai/rosetta/blob/main/docs/RFCs/Rosetta%20v3.0.0%20Core%20Spine%20Specification.md)
- Existing executable-world boundary: [#1517](https://github.com/entif-ai/rosetta/issues/1517)

Paper-grounded findings that constrain the design:

- Intersection identity is linearly decodable and causally effective; teleportation succeeds for 99.3% of testable intersections in at least one scene.
- The model uses recent position features: removing past-position information reduces late-ride current-position decoding from 99.8% to 14.8%, while preserving progressively longer windows restores it.
- A two-dimensional goal compass predicts goal bearing and causally steers motion.
- Intersection features are superposed in a comparatively small subspace and cluster more by legal-move affordance than geographic position.
- Out-of-distribution failures are associated with weakened true-position writes and increased interference, not necessarily destruction of the learned map.
- Successful causal repair is evidence about the tested mechanism under the declared intervention, not proof of a complete or uniquely correct world model.
- The code separates experiments for intersections, connectivity, tracking, compass, stopping, stress, dynamics, failures, detours, compression, and cross-checkpoint comparison. It warns that model/run artifacts and coverage metadata must remain paired and that distinct compression benchmarks are not interchangeable.

## 3. Why this belongs in Rosetta

The Core Spine already provides content-addressed Tiles and the Run/Action/ToolCall/Observation/Evaluation/Receipt sequence, with explicit provenance and uncertainty. This research supplies a concrete mechanistic-audit workload that pressures those boundaries:

- a high decoding score is not a causal-use claim;
- causal use of one feature is not a complete world-model claim;
- an internal decoded state is not an external-world observation;
- a sampled illegal action is not itself a diagnosis of the latent failure;
- an intervention that repairs behavior does not authorize runtime steering;
- a receipt that an experiment executed does not establish semantic or causal truth.

[#1517](https://github.com/entif-ai/rosetta/issues/1517) owns external physical evidence, executable hypotheses, simulations, and counterfactual-versus-history boundaries. This issue should consume those general distinctions where appropriate, but adds a different surface: **mechanistic evidence about a model’s internal state and causal use of learned representations**.

## 4. Concept and primitive crosswalk

| Paper/code construct | Proposed representation | Mapping | Required boundary |
|---|---|---:|---|
| TaxiGPT checkpoint and tokenizer | immutable ModelProfile and Artifact refs with digest | exact | checkpoint identity is not capability |
| Manhattan graph, valid turns, shortest paths | EnvironmentManifest Observation refs | exact | benchmark environment is not the learned representation |
| Held-out ride/prefix | Run input Observation plus selection provenance | exact | selected sample is not population evidence |
| Residual activation at layer/token | external payload ref plus ActivationObservation metadata | partial | activation is internal measurement, not semantic interpretation |
| Difference-of-means intersection direction | FeatureArtifact with extraction profile and coverage refs | exact | decodability is not causal use |
| Position subspace / 244-dimensional basis | SubspaceArtifact with layer, basis, fit split, and normalization | exact | basis identity and coverage must travel together |
| Linear coordinate readout | ProbeProfile, ToolCall, and Evaluation | exact | high R² does not prove the model uses geographic coordinates |
| Minimal teleportation / feature injection | InterventionAction with target, layer, strength, control, and seed | exact | intervention success is scoped to tested scenes |
| Rolling look-back ablation | InterventionAction plus StateTrackingEvaluation | exact | ablation effect does not uniquely identify mechanism |
| Goal compass | FeatureArtifact plus probe and steering Evaluations | partial | readable direction and used direction remain separately evidenced |
| Superposition/affordance packing | RepresentationStructureEvaluation | partial | clustering similarity is not an ontology or truth claim |
| Stress prefix | StressCase Observation with OOD rationale | exact | OOD membership and failure are separate claims |
| Illegal-move probability/output | BehaviorObservation | exact | behavior is not a latent-state diagnosis |
| Fatal/silent/full-corruption/give-up rule | FailureClassifierProfile plus Evaluation | exact | rule/threshold version must be preserved |
| Write restoration/noise clearing | RecoveryIntervention and RecoveryEvaluation | exact | recovery is not deployment authority |
| Cross-checkpoint indicator table | ComparisonEvaluation with comparable-profile refs | exact | incompatible samplers/scorers must not be merged |
| Figure/table artifact | Projection with source Evaluation refs | exact | a projection is not the canonical evidence |
| Experiment completion | Receipt | exact | receipt attests execution/integrity, not model-world correctness |

## 5. Recommended collaboration artifact

Use a **SchemaPack with an experimental interoperability profile**, not a Core extension.

Proposed root:

`packs/schema-pack-mechanistic-world-model-audit/`

Proposed profile:

`mechanistic-world-model-audit-v1`

Rationale:

- the paper and repository expose stable experiment families, artifact identities, causal interventions, metrics, and failure classes;
- multiple mechanistic-interpretability toolchains could emit the same public records without standardizing private extraction code;
- a pack can carry schemas, vocabulary, adapter examples, fixtures, and conformance tests while keeping Core stable;
- the profile can explicitly import or reference #1517 for environment and counterfactual semantics without conflating external-world hypotheses with internal model-state evidence.

## 6. Proposed architecture

1. **Manifest and identity layer**
   - Pin paper/version, source repository commit, code license, checkpoint digest, tokenizer/data manifests, environment graph, experiment configuration, random seeds, hardware/runtime summary, and artifact root.
2. **Mechanistic evidence layer**
   - Represent activation payloads by integrity reference; emit feature, subspace, probe, and coverage artifacts with fit/evaluation split identity.
3. **Intervention layer**
   - Record intervention target, layer/token scope, edit vector or external payload ref, strength, control condition, baseline state, and expected causal question.
4. **Behavior layer**
   - Record pre/post action probabilities, selected action, legality, trajectory continuation, stopping behavior, and recovery without interpreting these as internal state.
5. **Diagnosis layer**
   - Apply a versioned failure classifier to mechanistic and behavioral evidence; allow `unclassified` and competing diagnoses.
6. **Evaluation layer**
   - Emit decoding, causal-use, state-tracking, navigation, representation-structure, failure, recovery, and cross-checkpoint Evaluations.
7. **Reliance layer**
   - Mint receipts and bounded reliance records that name exactly which claim is supported, under which profile, sample scope, controls, and caveats.

Large tensors, centroid matrices, cached activations, probes, and trajectory arrays remain external typed payloads addressed by digest; the graph carries identity, schema, provenance, scope, and evaluation links.

## 7. Translation and interoperability design

Define a minimal adapter boundary:

```ts
interface MechanisticWorldModelAuditAdapter {
  describeRun(input: NativeRunRef): Promise<AuditRunManifest>;
  projectFeature(input: NativeFeatureRef): Promise<FeatureArtifact>;
  projectIntervention(input: NativeInterventionRef): Promise<InterventionRecord>;
  projectBehavior(input: NativeBehaviorRef): Promise<BehaviorObservation>;
  projectEvaluation(input: NativeEvaluationRef): Promise<MechanisticEvaluation>;
}
```

A conforming adapter must preserve:

- native artifact identity and digest;
- source repository/ref and code-version identity;
- checkpoint, tokenizer, dataset, split, and environment identity;
- layer, token position, representation space, normalization, and extraction procedure;
- fit-set versus evaluation-set boundaries;
- intervention target, control, strength, and seed;
- raw measured outputs separately from diagnostic interpretation;
- failure-classifier version, ordered rule, thresholds, and `unclassified` state;
- uncertainty, sample coverage, exclusions, and negative results;
- exact metric definition and aggregation;
- parent/child lineage for derived artifacts and projections.

Round-trip expectation: serialize a selected TaxiGPT experiment run, restore its profile and external payload refs, and reproduce the same classification/evaluation result within declared tolerance. The profile does not require byte-identical floating-point output across hardware unless a fixture declares that stronger condition.

## 8. Repository and implementation impact

Suggested files:

- `packs/schema-pack-mechanistic-world-model-audit/pack.yaml`
- `packs/schema-pack-mechanistic-world-model-audit/README.md`
- `packs/schema-pack-mechanistic-world-model-audit/vocab/mechanistic-world-model-audit.v1.yaml`
- `packs/schema-pack-mechanistic-world-model-audit/schemas/audit-run-manifest.schema.json`
- `packs/schema-pack-mechanistic-world-model-audit/schemas/feature-artifact.schema.json`
- `packs/schema-pack-mechanistic-world-model-audit/schemas/intervention-record.schema.json`
- `packs/schema-pack-mechanistic-world-model-audit/schemas/behavior-observation.schema.json`
- `packs/schema-pack-mechanistic-world-model-audit/schemas/mechanistic-evaluation.schema.json`
- `packs/schema-pack-mechanistic-world-model-audit/schemas/failure-classification.schema.json`
- `packs/schema-pack-mechanistic-world-model-audit/schemas/reliance-receipt.schema.json`
- `packs/schema-pack-mechanistic-world-model-audit/fixtures/taxigpt/`
- `packs/schema-pack-mechanistic-world-model-audit/adapters/world_modeling_reference.py`
- `packs/schema-pack-mechanistic-world-model-audit/tests/`
- pack-suite index and documentation cross-links.

Reuse existing Core schema references, source/provenance families, receipt closure, and external-payload conventions. Do not copy the paper repository into this repository.

## 9. Phased engineering plan

**Phase 0: evidence lock and profile ADR**

- Pin paper v1, project page, source repository commit, MIT license, and relevant Core/#1517 versions.
- Record all paper claims used by fixtures and distinguish reported results from proposed schema choices.
- Decide whether the profile is a child consumer of #1517 or only cross-references it.

**Phase 1: vocabulary and schemas**

- Define artifact kinds, evaluation types, controlled failure labels, claim scopes, and prohibited inferences.
- Implement JSON Schemas and semantic constraints.
- Define external-array payload addressing and coverage metadata.

**Phase 2: clean-room reference adapter**

- Map one run each from `intersections`, `tracking`, `compass`, and `failures`.
- Read existing outputs; do not reimplement model internals initially.
- Preserve native command, config, artifact root, and source-file provenance.

**Phase 3: conformance fixtures**

- Add positive fixtures for decoding, causal teleportation, history ablation, compass steering, failure classification, and repair.
- Add negative fixtures that intentionally collapse distinct claims or mix incompatible runs/metrics.

**Phase 4: validation and documentation**

- Add schema validation, referential-integrity checks, lineage closure, failure-rule replay, and prohibited-inference linting.
- Document adapter semantics, lossiness, and non-goals.

**Phase 5: researcher review and generalization**

- Invite the authors to review the crosswalk, especially feature/subspace identity, coverage metadata, causal claims, and failure categories.
- Only after that review, test a second model/environment or mechanistic toolchain.

## 10. Detailed task checklist

- [ ] Create a profile ADR defining scope and relationship to #1517.
- [ ] Pin source versions and license evidence.
- [ ] Define `AuditRunManifest` and external payload integrity conventions.
- [ ] Define model/checkpoint/tokenizer/data/environment identity fields.
- [ ] Define `FeatureArtifact`, `SubspaceArtifact`, and `ProbeProfile`.
- [ ] Define fit/evaluation split and coverage metadata.
- [ ] Define `InterventionRecord` including controls and target scope.
- [ ] Define `BehaviorObservation` separately from diagnosis.
- [ ] Define mechanistic Evaluation subtypes and uncertainty fields.
- [ ] Encode the ordered TaxiGPT failure-classification profile, including `unclassified`.
- [ ] Define recovery Evaluation and bounded reliance Receipt.
- [ ] Implement a clean-room projection adapter over released outputs.
- [ ] Add fixtures for intersections, tracking, compass, stress/failures, and recovery.
- [ ] Add cross-run and incompatible-metric negative fixtures.
- [ ] Add validation for missing checkpoint/data/code/config identities.
- [ ] Add validation preventing readout success from being asserted as causal use.
- [ ] Add validation preventing intervention success from becoming deployment authority.
- [ ] Add validation preventing receipt success from becoming semantic truth.
- [ ] Add pack documentation, examples, and pack-suite index entry.
- [ ] Record author-review questions and disposition without implying endorsement.

## 11. Acceptance criteria and verification

- [ ] Every projected evaluation resolves to a paper/code source, run manifest, checkpoint, dataset/split, configuration, and metric profile.
- [ ] Activation measurement, feature interpretation, decoded state, causal intervention, behavior, diagnosis, and reliance are distinct linked artifacts.
- [ ] Fit and evaluation data cannot be silently identical where the declared method requires separation.
- [ ] Feature/subspace artifacts retain layer, token scope, normalization, extraction method, coverage, and payload digest.
- [ ] Interventions retain target, control, strength, seed, and before/after observations.
- [ ] The ordered failure classifier reproduces the declared class for all TaxiGPT conformance fixtures.
- [ ] `unclassified`, ambiguous, and competing-diagnosis states validate.
- [ ] A successful decoder cannot satisfy a causal-use claim without intervention evidence.
- [ ] A successful repair cannot satisfy a general safety or deployment-authorization claim.
- [ ] A Receipt cannot satisfy world-model correctness or causal completeness.
- [ ] Cross-checkpoint comparisons reject mismatched datasets, samplers, scorers, or profile versions unless an explicit translation Evaluation exists.
- [ ] External tensor/trajectory payloads verify by digest without requiring them in graph nodes.
- [ ] Round-trip projection reproduces classification/evaluation within declared tolerance.
- [ ] Negative fixtures fail for mixed artifact roots, missing coverage metadata, silent metric substitution, and collapsed observation/interpretation claims.
- [ ] CI validates schemas, references, fixtures, and prohibited-inference rules.

## 12. Risks, constraints, and licensing

- **External dependencies:** TaxiGPT datasets/checkpoints follow upstream instructions and are not all bundled; small-model weights and training code are also not bundled.
- **Compute:** many measurements require a GPU; reduced budgets check execution but are not substitutes for reported measurements.
- **Floating point:** exact numerical reproduction may depend on device/library versions.
- **Mechanistic underdetermination:** correlated or decodable features need not be uniquely causal or exhaustive.
- **Intervention artifacts:** feature directions can be method- and layer-specific; the profile must not promote them to stable semantic identities without evidence.
- **Benchmark scope:** Manhattan random-walk navigation is a controlled domain, not a demonstration about all transformer world models.
- **Threshold dependence:** failure labels depend on ordered rules and thresholds that must be versioned.
- **Safety:** this profile records intervention evidence; it grants no authority to modify a deployed model.
- **License:** the released code is MIT licensed; upstream checkpoints/data retain their own terms. Preserve notices and never imply those assets are relicensed by the pack.
- **Privacy:** fixtures should use public benchmark artifacts and avoid private activation data.

## 13. Open questions for the researchers

- Which native output objects are the most stable interoperability surface: centroids/subspaces, per-scene traces, classification tables, or experiment summaries?
- Which coverage metadata must travel with an intersection-feature artifact to prevent invalid reuse?
- Should the four failure categories be represented as exclusive outputs of one ordered classifier, or also as independently scored hypotheses?
- Which intervention results most strongly justify “causally used” rather than merely “causally correlated”?
- How should the profile distinguish helpful affordance packing from harmful interference without turning a descriptive metric into a normative threshold?
- Which controls are indispensable when moving from TaxiGPT to other transformers or learned environments?
- Are there results currently saved only as figure/table intermediates that should become first-class machine-readable artifacts?
- What reproducibility tolerance is appropriate across hardware and library versions?

## 14. Non-goals

- Defining a universal theory or ontology of world models.
- Claiming that TaxiGPT results generalize to frontier language models.
- Treating internal activations as direct observations of external reality.
- Standardizing feature extraction, probe fitting, sparse autoencoding, or causal-edit algorithms.
- Selecting safety thresholds or authorizing runtime interventions.
- Copying or vendoring upstream datasets/checkpoints without their licenses.
- Replacing #1517’s external executable-world contract.
- Modifying the Core Spine.
- Claiming author endorsement.

## 15. Collaboration path

1. Publish the profile ADR and initial concept crosswalk.
2. Ask the researchers to review the boundaries among decodability, causal use, localization, navigation, and failure diagnosis.
3. Project a small public fixture set from the released repository without changing its native formats.
4. Compare the projected records against the authors’ own tables/plots and document any semantic loss.
5. Iterate on vocabulary and fixture scope before proposing broader mechanistic-tool interoperability.
6. Keep collaboration optional, public, technically bounded, and strictly non-commercial.

## 16. Source references

- https://arxiv.org/abs/2609.21748
- https://bepierre.github.io/world-modeling/
- https://github.com/bepierre/world-modeling
- https://github.com/bepierre/world-modeling/blob/main/experiments/README.md
- https://entif.ai/tags/research/2026/09/10/babel-fish-alien-mind/
- https://entif.ai/tags/research/2026/09/09/rosetta-pasigraphy-protocol/
- https://github.com/entif-ai/rosetta/blob/main/docs/RFCs/Rosetta%20v3.0.0%20Core%20Spine%20Specification.md
- https://github.com/entif-ai/rosetta/issues/1517
